### PR TITLE
feat: derive a market-implied point forecast from bucket order books

### PR DESCRIPTION
**Pull request by vinarmani**
Prediction markets price RANGES. A five-bucket EPS market says "34% chance EPS lands between $2.06 and $2.21"; it never says "EPS will be $2.14". This adds the inverse: given the order books across every bucket of one market, produce the single number they collectively imply and the band around it.

## What

- `trufnetwork_sdk_py.forecast` — the algorithm, as pure functions with no I/O: `forecast_from_depth` (preferred), `forecast_from_buckets`, and the `BucketDepth` / `BookLevel` / `BucketBook` / `MarketForecast` types.
- `TNClient.get_market_forecast(query_ids)` — fetches the books for a market's buckets and returns the forecast.
- `trufnetwork_sdk_py.market_buckets.bucket_bounds_from_market_data` — turns decoded market data into a bucket's `(lower, upper)` bounds.
- API reference section covering all of the above.

## Why

The algorithm is validated against live mainnet books in truflation/prediction-bots, where it drives quoting. Callers of this SDK — the indexer, dashboards, anything reading a market — otherwise have to reimplement it or go without.

## How

`forecast.py` is a VERBATIM copy of
truflation/prediction-bots@21fe4f3 (`market-maker-bot/src/market_maker_bot/ forecast.py`), carrying only an added provenance header. That is deliberate. The module is numerics that were tuned against real books and it is still moving fast — three substantive revisions in the past week — so re-syncing has to stay a `diff` rather than a re-transcription. Style was left un-adapted for the same reason. SDK-specific glue lives in `market_buckets.py` instead of being added to the mirrored file, so the diff stays clean.

`get_market_forecast` reads both the YES and NO ladders per bucket. On this venue a resting BUY NO at p is hittable by a BUY YES at 100-p (mint match), so NO liquidity is executable YES liquidity; ignoring it discards real quotes. That costs two order-book reads per bucket.

## Notes for callers

`margin_of_error` is half the P10..P90 band — how uncertain the OUTCOME is, not how tightly the book pins the estimate. It is large by design (~0.24 against a value of ~2.14 on a live EPS market). The `value_basis` / `p10_basis` / `p90_basis` flags say whether a number was read between real strikes or supplied by the tail model for an open-ended outer bucket.

## Testing

57 offline tests, ported from upstream alongside the module, plus SDK-side tests for bounds derivation and client assembly.

`tests/test_forecast_live.py` adds an env-gated check against a real network (`TN_LIVE_NODE_URL`), skipped by default. Unit tests feed the algorithm hand-written numbers and so cannot catch the SDK reading the chain wrongly; its central assertion cross-checks `get_order_book`'s negative-price bid convention against `get_best_prices`, which reports the same bid positive. If either convention changes, that fails loudly instead of silently inverting every one-sided bucket.

Verified end-to-end against mainnet: the discovered NVDA EPS market (query_ids 419-423) forecasts 2.1363 with a 1.9058..2.3792 band, agreeing with an independent top-of-book read to within 0.002.

## Summary by CodeRabbit

* **New Features**
  * Added market-implied forecasting from prediction-market quotes and order-book depth.
  * Added forecast values, uncertainty bands, bucket probabilities, confidence indicators, warnings, and serialization.
  * Added client support for retrieving forecasts across market buckets.
  * Added utilities for deriving bucket bounds from market metadata.
* **Documentation**
  * Added API documentation and usage examples for forecasting.
* **Tests**
  * Added coverage for quote-based, depth-based, live-market, and edge-case forecasting scenarios.

### CODE REVIEW

File: tests/test_forecast_live.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                assert 0 < level.price < 100, (
+                    f"price {level.price} outside the 1-99 cent convention"
+                )
+                assert level.size > 0
+
+            if bids and best["best_bid"] is not None:
+                assert max(l.price for l in bids) == best["best_bid"]
+                checked += 1
+            if asks and best["best_ask"] is not None:
+                assert min(l.price for l in asks) == best["best_ask"]
```
[the remark this answers is not in the document]
Comment (vinarmani) [reply tests/test_forecast_live.py]: There is no Ruff in this project: no [tool.ruff] in pyproject.toml, no ruff.toml/.ruff.toml/setup.cfg/tox.ini, and no lint job in .github/

**Review by MicBun — CHANGES_REQUESTED** [review]
Requesting changes on two items. The rest checks out, including the verbatim claim: `forecast.py` differs from `prediction-bots@21fe4f3` only by the 9-line header, so the resync-by-diff strategy holds.

### The identity check is inert here but active in the other two SDKs

`client.py` raises only when `timestamp` is present and None, and the comment explains why: the committed bindings never emit the field, so today it is absent for every bucket and the check ignores it. sdk-go's `requireQueryTime` and the sdk-js equivalent both have the field and reject.

The effect is that two buckets sharing a settle time but observing the stream at different points get accepted here and rejected by the other two SDKs. Same input, different answer, depending on which SDK the caller picked.

Nothing forces the binding rebuild, and the description does not mention the coupling. Either land this after sdk-go #197 with the bindings rebuilt, or make the absent-key case fail the same way the None case does so the behaviour stops depending on a build artifact.

CodeRabbit's comment about the binding omitting `timestamp` and `frozen_at` covers part of this and is still unanswered.

### The headline example raises TypeError

`docs/api-reference.md:1497` formats `forecast.p10` with `:.4f`, and the field table below types it `float | None`. On the discrete fallback both percentiles are None, so the first example a reader copies crashes.

Worth fixing in the same pass: the example prints `.value` without checking `value_basis`. The tail model is unbounded when mass sits in the open-ended outer bucket. A five-bucket book struck 4.04 to 4.92 with 90% of the mass above the top strike returns 8.09 with a p90 of 16.79:

```python
bounds = [4.04, 4.33, 4.62, 4.92]
probs  = [0.01, 0.02, 0.02, 0.05, 0.90]
_rank_quantile(bounds, probs, 0.50)   # (8.094, 'tail')
_rank_quantile(bounds, probs, 0.90)   # (16.785, 'tail')
```

The basis flags and the warning both say so, which is the design. The example teaches callers to skip them.

**Review by MicBun — APPROVED** [review]
I will make a follow up PR after this
